lib/manageiq/api/common/application_controller_mixins/exception_handling.rb

```diff
@@ -0,0 +1,43 @@
+module ManageIQ
+  module API
+    module Common
+      module ApplicationControllerMixins
+        module ExceptionHandling
+          DEFAULT_ERROR_CODE = 400
+
+          def self.included(other)
+            other.rescue_from(StandardError, RuntimeError) do |exception|
+              errors, status = collected_errors(exception)
+
+              render :json => errors, :status => status
+            end
+          end
+
+          def collected_errors(exception)
+            errors = []
+            top_level_exception = exception
+
+            until exception.nil?
+              code = exception.respond_to?(:code) ? exception.code : error_code_from_class(exception)
+              errors << {:status => code, :detail => exception.class.to_s + ": " + exception.message}
+              exception = exception.cause
+            end
+
+            # Overwrite the first error since that one most likely doen't have a code
+            errors.first[:status] = error_code_from_class(top_level_exception)
+
+            [{:errors => errors}, error_code_from_class(top_level_exception)]
+          end
+
+          def error_code_from_class(exception)
+            if ActionDispatch::ExceptionWrapper.rescue_responses.key?(exception.class.to_s)
+              ActionDispatch::ExceptionWrapper.rescue_responses[exception.class.to_s]
+            else
+              DEFAULT_ERROR_CODE
+            end
```

Comment:
Can this be simplified to `ActionDispatch::ExceptionWrapper.rescue_responses[exception.class.to_s] || DEFAULT_ERROR_CODE` ?

Reply:
See thats what I tried initially - but `ActionDispatch::ExceptionWrapper.rescue_responses` is a hash thats created with a default value of `:service_unavailable` so if the key doesn't exist it never falls back to the default code. 

src: https://github.com/rails/rails/blob/a3d54d2afe557b243d6b18f9adf119f60ffb8242/actionpack/lib/action_dispatch/middleware/exception_wrapper.rb#L8